src/hooks/useAsync.ts

```diff
@@ -33,10 +27,28 @@ const useSafeDispatch = (dispatch: any) => {
     );
 };
 
-const useAsync = (initialState?: any) => {
+export interface UseAsyncResponse<T, U> {
+    data: U extends undefined ? T | null : T;
+    error: any;
+    isError: boolean;
+    isIdle: boolean;
+    isLoading: boolean;
+    isSuccess: boolean;
+    reset: () => any;
+    run: (promise: Promise<any>, dataFetcher?: Function | undefined) => any;
+    setData: (setDataResponse: T) => void;
+    setError: (setErrorResponse: any) => any;
+    status: States;
+}
+
+function useAsync<T>(): UseAsyncResponse<T, undefined>;
+function useAsync<T>(initialData: T): UseAsyncResponse<T, T>;
+function useAsync<T>(initialData?: T): UseAsyncResponse<T, undefined> {
```

Comment:
Easiest way I would figure out how to type that if there is initial data then data should not be nullable.